modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/GridCacheSetImpl.java

```diff
@@ -553,12 +623,20 @@ GridCacheContext context() {
         return ctx;
     }
 
+    /**
+     * @return {@code True} If this instance is using shared cache for store set data.
+     */
+    boolean sharedCacheMode() {
+        return sharedCacheMode;
+    }
+
     /**
      * @param item Set item.
      * @return Item key.
      */
     private SetItemKey itemKey(Object item) {
-        return collocated ? new CollocatedSetItemKey(name, id, item) : new GridCacheSetItemKey(id, item);
+        return collocated ? new CollocatedSetItemKey(name, id, item) :
+            new GridCacheSetItemKey(sharedCacheMode ? id : null, item);
```

Comment:
may be add sharedCacheMode flag into GridCacheSetItemKey ?

Reply:
No -this will increase set data size (each key).